rust/cubesql/cubesql/src/telemetry/mod.rs

```diff
@@ -134,18 +73,86 @@ impl Log for ReportingLogger {
     }
 
     fn log<'a>(&self, record: &Record<'a>) {
-        if let Level::Error = record.metadata().level() {
-            track_event_spawn(
-                "Cube SQL Error".to_string(),
-                vec![("error".to_string(), format!("{}", record.args()))]
-                    .into_iter()
-                    .collect(),
-            )
+        match (record.metadata().level(), self.need_to_report) {
+            // reporting errors only
+            (Level::Error, true) => {
+                report(
+                    "Cube SQL Error".to_string(),
+                    HashMap::from([("error".to_string(), format!("{}", record.args()))]),
+                    Level::Error,
+                );
+            }
+            _ => self.logger.log(record),
         }
-        self.logger.log(record)
     }
 
     fn flush(&self) {
         self.logger.flush()
     }
 }
+
+pub trait ContextLogger: Send + Sync + Debug {
+    fn error(&self, message: &str);
+}
+
+#[derive(Debug)]
+pub struct SessionLogger {
+    session_state: Arc<SessionState>,
+}
+
+impl SessionLogger {
+    pub fn new(session_state: Arc<SessionState>) -> SessionLogger {
+        Self { session_state }
+    }
+
+    fn log(&self, target: &str, props: HashMap<String, String>, level: Level) {
+        // TODO: MySQL app_name
+        let app_name =
+            if let Some(name) = self.session_state.all_variables().get("application_name") {
+                name.value.to_string()
+            } else {
+                "NULL".to_string()
+            };
+        let protocol = self.session_state.protocol.to_string();
+
+        let mut meta_fields = props;
+        meta_fields.insert("appName".to_string(), app_name);
+        meta_fields.insert("protocol".to_string(), protocol);
+
+        if report(target.to_string(), meta_fields.clone(), level) == false {
+            log::log!(target: target, level, "{:?}", meta_fields);
+        }
+    }
+}
+
+impl ContextLogger for SessionLogger {
+    fn error(&self, message: &str) {
+        self.log(
+            "Cube SQL Error",
+            HashMap::from([("error".to_string(), message.to_string())]),
+            Level::Error,
+        );
+    }
+}
+
+fn report(event: String, properties: HashMap<String, String>, level: Level) -> bool {
+    let guard = REPORTER
+        .read()
+        .expect("failed to unlock REPORTER for reading");
+    if !guard.is_active() {
+        return false;
+    }
+
+    let mut properties = properties;
+    properties.insert("id".to_string(), nanoid!(16));
+    properties.insert("anonymousId".to_string(), "cubesql".to_string());
```

Comment:
Cube.js already knows how to generate it; let's drop this line and allow Cube.js to handle it.